File: python/39.py
class Solution(object):
        
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res=[]
        candidates.sort(reverse=True)
        can=candidates
        for i in candidates:
            if target-i<0:
                continue
            mid=self.combinationSum([j for j in candidates if j<=i],target-i)
            if target-i==0:
                res.append([i])
                
            if mid!=[]:
                for al in mid :
                    al.append(i)
                    if al not in res:
                        res.append(al)
                
        return res
            
            
class Solution:
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        candidates.sort()
        
        def dfs(target, index, path):
            if target < 0:
                return  # backtracking
            if target == 0:
                res.append(path)
                return 
            for i in range(index, len(candidates)):
                
                dfs(target-candidates[i], i, path+[candidates[i]])
        
        dfs(target, 0, [])
        return res

File: java/BFS&DFS/39.java
class Solution {
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {

        if (target == 0)
            return res;
        Arrays.sort(candidates);
        List<Integer> list = new ArrayList<>();
        _bc(candidates, 0, target, list);
        return res;

    }

    private void _bc(int[] candidates, int ind, int target, List<Integer> ans) {
        if (target < 0)
            return;
        if (target == 0) {
            res.add(new ArrayList<Integer>(ans));
            return;
        }
        for (int i = ind; i < candidates.length; i++) {
            if (i > 0 && candidates[i] == candidates[i - 1])
                continue;
            ans.add(candidates[i]);
            _bc(candidates, i, target - candidates[i], ans);
            ans.remove(ans.size() - 1);

        }
    }

}
